packages/preview/src/browser/markdown/markdown-preview-handler.ts

```diff
@@ -0,0 +1,109 @@
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+import { inject, injectable } from "inversify";
+import { ResourceProvider } from "@theia/core";
+import { PreviewHandler } from '../preview-handler';
+import URI from "@theia/core/lib/common/uri";
+
+import * as hljs from 'highlight.js';
+import * as markdownit from 'markdown-it';
+
+import '../../../src/browser/markdown/style/index.css';
+
+@injectable()
+export class MarkdownPreviewHandler implements PreviewHandler {
+
+    @inject(ResourceProvider)
+    protected readonly resourceProvider: ResourceProvider;
+
+    readonly iconClass: string = 'markdown-icon file-icon';
+    readonly contentClass: string = 'markdown-preview';
+
+    canHandle(uri: URI): number {
+        return uri.path.ext === '.md' ? 500 : 0;
+    }
+
+    renderHTML(content: string): string {
+        return this.getEngine().render(content);
+    }
+
+    findElementForSourceLine(sourceLine: number, renderedNode: HTMLElement): HTMLElement | undefined {
+        const markedElements = renderedNode.getElementsByClassName('line');
+        let matchedElement: HTMLElement | undefined;
+        for (let i = 0; i < markedElements.length; i++) {
+            const element = markedElements[i];
+            const line = Number.parseInt(element.getAttribute('data-line') || '0');
+            if (line > sourceLine) {
+                break;
+            }
+            matchedElement = element as HTMLElement;
+        }
+        return matchedElement;
+    }
+
+    getSourceLineForOffset(content: HTMLElement, offset: number): number | undefined {
+        const filter: NodeFilter = {
+            acceptNode: (node: Node) => {
+                if (node instanceof HTMLElement) {
+                    if (node.classList.contains('line')) {
+                        return NodeFilter.FILTER_ACCEPT;
+                    }
+                    return NodeFilter.FILTER_SKIP;
+                }
+                return NodeFilter.FILTER_REJECT;
+            }
+        };
+        const treeWalker = document.createTreeWalker(content, NodeFilter.SHOW_ELEMENT, filter, false);
+        let lastElement: HTMLElement | undefined = undefined;
+        while (treeWalker.nextNode()) {
+            const current = treeWalker.currentNode as HTMLElement;
+            const offsetTop = current.offsetTop;
+            if (offsetTop > offset) {
+                break;
+            }
+            lastElement = current;
+        }
+        if (!lastElement) {
+            return undefined;
+        }
+        const line = Number.parseInt(lastElement.getAttribute('data-line') || '0');
+        return line;
+    }
+
+    protected engine: markdownit.MarkdownIt | undefined;
+    protected getEngine(): markdownit.MarkdownIt {
+        if (!this.engine) {
+            const engine: markdownit.MarkdownIt = this.engine = markdownit({
```

Comment:
yes, I also like the idea of renderer extensions. 'sounds familiar to me ;-)

Reply:
It would be nice to decouple the markdown renderer to be able to use it outside of the previewer.

Reply:
@akosyakov, there are several options how to extract the markdownit engine from this preview handler. it may depend on other requirements. 

configuring of plugins and renderers happens at construction time. while additional renderers could be controlled through render method's options parameter, I see no way to disable plugins. OTOH we could have a factory for markdownit engines with options, e.g. `createMarkdownEngine(options: MarkdownEngineOptions) : MarkdownIt`

Reply:
Let's do that in another step.